### PR TITLE
Refactor decl unit testing for easy validation testing

### PR DESCRIPTION
**Pull request by out-of-phaze**
## Description of changes
Decl UID uniqueness checking was moved to `unique_tests.dm` and now takes advantage of the helpers for uniqueness testing.

Replaces the `mandatory_uid_types` list and `crash_on_abstract_init` var with two bitflags on the `decl_flags` var. Crashing on abstract init is now the default behavior, exempted only by `/decl/hierarchy` at the moment.

Adds a new proc, `/decl/proc/validate()`, used in unit testing to ensure that a decl's settings are sane. By default, it handles enforcing that UIDs are either text or null (or just text, on types with `DECL_FLAG_MANDATORY_UID`). It returns a list of informative failure reasons as strings, and is printed along with the decl's type in unit testing.

`/decl/trait` unit tests have been moved into `/decl/trait/validate`, `/decl/species/validate`, and the uniqueness tests file.

Adds validation to cocktails to ensure that `ratios` is non-empty.

Marks several base decl types as abstract (since they're all uninitialised). See: culture descriptor, culture, faction, location, religion.

- [x] Convert materials unit tests to `/decl/material/validate()`
- [x] Convert special role unit tests to `/decl/special_role/validate()`
- [x] Convert submap_archetype unit tests to uniqueness test and `/decl/submap_archetype/validate()`.
- [x] Convert security_level unit tests to `/decl/security_level/validate()`.
- [x] Convert aspect unit tests to uniqueness tests and `/decl/aspect/validate()`. MUST LEAVE `aspects_shall_not_have_circular_references` AS-IS
- [x] Convert currency unit tests to `/decl/currency/validate()`.
- [x] Convert music_track tests to `/decl/music_track/validate()`.
- [x] Convert surgery_step tests to `/decl/surgery_step/validate()`.
- LEAVE `chemical_reaction` UNIT TESTS ALONE

## Why and what will this PR improve
An easy framework to test decls without having to add and include an entirely new unit tests file and write boilerplate code for it is pretty dang neat.

## Authorship
Me, with @MistakeNot4829's assistance.

### CODE REVIEW

**Comment by out-of-phaze** [comment]
Relies on #2625.

**Comment by out-of-phaze** [comment]
Should be ready as soon as the recent stable fixes get merged into dev.

**Comment by out-of-phaze** [comment]
Ready for review once #2797 is in.

File: code/modules/materials/_materials.dm [hunk 1]
```diff
@@ -334,6 +335,67 @@ INITIALIZE_IMMEDIATE(/obj/effect/gas_overlay)
...
+	if(icon_reinf)
+		if(use_reinf_state)
+			if(!check_state_in_icon(use_reinf_state, icon_reinf))
+				. += "'[icon_reinf]' - missing reinf icon state '[use_reinf_state]'"
+		else
+			for(var/i = 0 to 7)
+				if(!check_state_in_icon("[i]", icon_reinf))
+					. += "'[icon_reinf]' - missing directional reinf icon state '[i]'"
+
+	if(length(color) != 7)
```
Comment (MistakeNot4892) [inline code/modules/materials/_materials.dm:395]: 9 is also valid (alpha channel)

**Comment by MistakeNot4892** [comment]
Other than the alpha comment, this looks good to me. Lots of room for expanding these out easily, nice PR.

File: code/modules/materials/_materials.dm (hunk 1 above)
Comment (out-of-phaze) [reply code/modules/materials/_materials.dm:395]: No, the original test existed to make it so they didn't have alpha channels. It's copied from `"CHEMISTRY: Reagents must have valid colors without alpha in color value"`.

**Review by MistakeNot4892 — APPROVED** [review]

**Pull request merged**